### PR TITLE
Refactor ColumnSelectorFactory; Rely on ColumnValueSelector's polymorphism

### LINKED ISSUE #4800: Remove duplicating polymorphism of ColumnValueSelectors
E. g. remove ColumnSelectorFactory.makeFloat/Double/Long/ObjectColumnSelector, and leave only one method "makeColumnValueSelector". Users which need different specializations could call corresponding methods on `ColumnValueSelector` itself: getFloat()/getDouble/getLong().

This change will allow to reduce code duplication in `IndexedDoubles`, `IndexedLongs` and `IndexedFloats`.

### PR DESCRIPTION
**Pull request by leventov**
Fixes #4800. Removes a lot of long/float/double/object code repetition across the project.

### CODE REVIEW

**Comment by leventov** [comment]
This PR fails on tests in `SchemaEvolutionTest`. According to them, numeric aggregation (e. g. longSum) on string dimension column should result zeros, currently with this PR it leads to exception. @himanshug @gianm what do you think it should do?

**Comment by himanshug** [comment]
@leventov IIRC, yes Druid tries to be as flexible as much possible and `longSum` works on string columns if they are parsable to long values or else assumed zeros.  typically aggregators try to accommodate  different column types if possible.  i think we should try and preserve the behavior.

**Comment by leventov** [comment]
@himanshug in the long term/general, what do you think about Druid to be "weakly typed", a-la MySQL, vs "strongly typed", a-la PostgreSQL?

**Comment by leventov** [comment]
#4888

**Comment by gianm** [comment]
@leventov, re:

> This PR fails on tests in SchemaEvolutionTest. According to them, numeric aggregation (e. g. longSum) on string dimension column should result to zeros, but currently with this PR it leads to exception. @himanshug @gianm what do you think it should do?

I commented on https://github.com/druid-io/druid/issues/4888#issuecomment-333610705 in more detail but wanted to add here that SchemaEvolutionTest's purpose is to make sure you can change types of columns for newer segments and have the query system handle that elegantly. It isn't flawless right now (for example you would probably expect a "sum" aggregator of a string column to parse the strings, rather than return zero) but it works well for dimensions.

> in the long term/general, what do you think about Druid to be "weakly typed", a-la MySQL, vs "strongly typed", a-la PostgreSQL?

I see it as being even weaker than MySQL on a table level. MySQL has a table-wide schema and I don't see Druid as ever requiring that.

**Comment by leventov** [comment]
Could someone please review this PR?

**Comment by himanshug** [comment]
i'll try and review it tomorrow.

**Comment by leventov** [comment]
@himanshug thank you

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 1]
```diff
@@ -118,53 +118,48 @@
...
-  /**
-   * @deprecated always throws {@link UnsupportedOperationException}
-   */
   @Deprecated
   @Override
   default float getFloat()
   {
-    throw new UnsupportedOperationException("DimensionSelector cannot be operated as numeric ColumnValueSelector");
+    // This is controversial, see https://github.com/druid-io/druid/issues/4888
+    return 0.0f;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: i thought we were gonna try and parse if possible.

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 2]
```diff
@@ -118,53 +118,48 @@
...
-  /**
-   * @deprecated always throws {@link UnsupportedOperationException}
-   */
   @Deprecated
   @Override
   default double getDouble()
   {
-    throw new UnsupportedOperationException("DimensionSelector cannot be operated as numeric ColumnValueSelector");
+    // This is controversial, see https://github.com/druid-io/druid/issues/4888
+    return 0.0;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: same here

File: processing/src/main/java/io/druid/segment/DimensionSelector.java [hunk 3]
```diff
@@ -118,53 +118,48 @@
...
-  /**
-   * @deprecated always throws {@link UnsupportedOperationException}
-   */
-  @Deprecated
-  @Override
-  default Object getObject()
-  {
-    throw new UnsupportedOperationException("DimensionSelector cannot be operated as object ColumnValueSelector");
+    // This is controversial, see https://github.com/druid-io/druid/issues/4888
+    return 0L;
```
Comment (himanshug) [inline processing/src/main/java/io/druid/segment/DimensionSelector.java]: same here

File: java-util/src/main/java/io/druid/java/util/common/StringUtils.java [hunk 4]
```diff
@@ -139,4 +139,27 @@ public static String toUpperCase(String s)
...
+
+  public static String removeChar(String s, char c)
+  {
+    for (int i = 0; i < s.length(); i++) {
+      if (s.charAt(i) == c) {
+        return removeChar(s, c, i);
+      }
+    }
+    return s;
+  }
```
[the remark this answers is not in the document]
Comment (leventov) [reply java-util/src/main/java/io/druid/java/util/common/StringUtils.java]: `''` is impossible in Java :)

File: processing/src/main/java/io/druid/segment/DimensionSelector.java (hunk 1 above)
Comment (leventov) [reply processing/src/main/java/io/druid/segment/DimensionSelector.java]: I preserved current behaviour. We can try to parse in a separate PR. Actually I tried to change this to parsing and it broke existing tests in `SchemaEvolutionTest`

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java [hunk 5]
```diff
@@ -122,12 +116,11 @@ public void reset(ColumnValueSelector selector)
...
-        SketchHolder other = ((ObjectColumnSelector<SketchHolder>) selector).getObject();
-        // SketchAggregatorFactory.combine() delegates to SketchHolder.combine() and it doesn't check for nulls, so we
-        // neither.
-        other.updateUnion(union);
-        combined.invalidateCache();
+        SketchHolder other = (SketchHolder) selector.getObject();
+        if (other != null) {
+          other.updateUnion(union);
+          combined.invalidateCache();
+        }
```
[the remark this answers is not in the document]
Comment (leventov) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java]: Yes, it is going to be used in later PRs in the project https://github.com/druid-io/druid/projects/2

File: processing/src/main/java/io/druid/segment/ColumnSelectorFactory.java [hunk 6]
```diff
@@ -30,12 +30,7 @@
 public interface ColumnSelectorFactory
 {
   DimensionSelector makeDimensionSelector(DimensionSpec dimensionSpec);
-  FloatColumnSelector makeFloatColumnSelector(String columnName);
-  LongColumnSelector makeLongColumnSelector(String columnName);
-  DoubleColumnSelector makeDoubleColumnSelector(String columnName);
-
-  @Nullable
-  ObjectColumnSelector makeObjectColumnSelector(String columnName);
+  ColumnValueSelector makeColumnValueSelector(String columnName);
```
[the remark this answers is not in the document]
Comment (leventov) [reply processing/src/main/java/io/druid/segment/ColumnSelectorFactory.java]: Do you have custom impls of `ColumnSelectorFactory`? This interface is not extended in any extensions in extensions-core or contrib

File: java-util/src/main/java/io/druid/java/util/common/StringUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (himanshug) [reply java-util/src/main/java/io/druid/java/util/common/StringUtils.java]: I meant using https://docs.oracle.com/javase/8/docs/api/java/lang/String.html#replace-char-char-

File: processing/src/main/java/io/druid/segment/ColumnSelectorFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (himanshug) [reply processing/src/main/java/io/druid/segment/ColumnSelectorFactory.java]: No-one is expected to extend `ColumnSelectorFactory` but it is used in custom extensions in implementations of  `AggregatorFactory.factorize/factorizedBuffered` which would be calling removed methods like `ColumnSelectorFactory.makeFloatColumnSelector(..)` etc. That is why `ColumnSelectorFactory` needs to have `@PublicApi` and not `@ExtensionPoint` annotation.

File: java-util/src/main/java/io/druid/java/util/common/StringUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (leventov) [reply java-util/src/main/java/io/druid/java/util/common/StringUtils.java]: `newChar` is char, `''` is not a char

File: api/src/main/java/io/druid/data/input/MapBasedRow.java [hunk 7]
```diff
@@ -123,51 +133,20 @@ public float getFloatMetric(String metric)
...
-    if (metricValue instanceof Number) {
-      return ((Number) metricValue).doubleValue();
-    } else if (metricValue instanceof String) {
-      try {
-        return Double.valueOf(((String) metricValue).replace(",", ""));
-      }
-      catch (Exception e) {
-        throw new ParseException(e, "Unable to parse metrics[%s], value[%s]", metric, metricValue);
+    if (metricValueString.charAt(0) == '+') {
+      if (metricValueString.length() > 1) {
```
[the remark this answers is not in the document]
Comment (leventov) [reply api/src/main/java/io/druid/data/input/MapBasedRow.java]: Fixed

File: processing/src/main/java/io/druid/segment/ColumnSelectorFactory.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (leventov) [reply processing/src/main/java/io/druid/segment/ColumnSelectorFactory.java]: Added `@PublicApi` to this interface and some others. So the next released version of Druid needs to be 0.12, I think.

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregator.java [hunk 8]
```diff
@@ -24,19 +24,24 @@
...
 public class SketchAggregator implements Aggregator
 {
-  private final ObjectColumnSelector selector;
-
+  private final ColumnValueSelector selector;
+  private final int size;
   private Union union;
 
-  public SketchAggregator(ObjectColumnSelector selector, int size)
+  public SketchAggregator(ColumnValueSelector selector, int size)
```
[the remark this answers is not in the document]
Comment (leventov) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregator.java]: Changed

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java [hunk 9]
```diff
@@ -66,24 +65,16 @@ public SketchAggregatorFactory(String name, String fieldName, Integer size, byte
...
   public Aggregator factorize(ColumnSelectorFactory metricFactory)
   {
-    ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(fieldName);
-    if (selector == null) {
-      return new EmptySketchAggregator();
-    } else {
-      return new SketchAggregator(selector, size);
-    }
+    ColumnValueSelector selector = metricFactory.makeColumnValueSelector(fieldName);
+    return new SketchAggregator(selector, size);
```
[the remark this answers is not in the document]
Comment (leventov) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java]: > null checks are removed, so in the new world it is guaranteed that for absent columns we get a non-null selector which would just return nulls/zeros from getXXX() methods ?

Yes, added some javadoc about this

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java [hunk 10]
```diff
@@ -34,13 +34,13 @@
 
 public class SketchBufferAggregator implements BufferAggregator
 {
-  private final ObjectColumnSelector selector;
+  private final ColumnValueSelector selector;
```
[the remark this answers is not in the document]
Comment (leventov) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchBufferAggregator.java]: Changed

File: processing/src/main/java/io/druid/query/aggregation/ObjectAggregateCombiner.java [hunk 11]
```diff
@@ -19,11 +19,18 @@
...
 /**
  * Specialization of {@link AggregateCombiner} for object aggregations.
  */
 public abstract class ObjectAggregateCombiner<T> implements AggregateCombiner<T>, ObjectColumnSelector<T>
 {
+
+  @Override
+  public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+  {
+    // Usually AggregateCombiner has nothing to inspect
```
[the remark this answers is not in the document]
Comment (leventov) [reply processing/src/main/java/io/druid/query/aggregation/ObjectAggregateCombiner.java]: Yes

File: extensions-core/stats/src/main/java/io/druid/query/aggregation/variance/VarianceAggregatorFactory.java [hunk 12]
```diff
@@ -116,20 +116,14 @@ public Aggregator factorize(ColumnSelectorFactory metricFactory)
   @Override
   public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
   {
-    ObjectColumnSelector selector = metricFactory.makeObjectColumnSelector(fieldName);
-    if (selector == null) {
+    ColumnValueSelector<?> selector = metricFactory.makeColumnValueSelector(fieldName);
+    if (selector instanceof NilColumnValueSelector) {
```
[the remark this answers is not in the document]
Comment (leventov) [reply extensions-core/stats/src/main/java/io/druid/query/aggregation/variance/VarianceAggregatorFactory.java]: Yes

File: processing/src/main/java/io/druid/query/aggregation/ObjectAggregateCombiner.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (leventov) [reply processing/src/main/java/io/druid/query/aggregation/ObjectAggregateCombiner.java]: Fixed

File: extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java [hunk 13]
```diff
@@ -99,8 +90,11 @@ public Object deserialize(Object object)
   }
 
   @Override
-  public Object combine(Object lhs, Object rhs)
+  public Object combine(Object lhs, @Nullable Object rhs)
```
[the remark this answers is not in the document]
Comment (leventov) [reply extensions-core/datasketches/src/main/java/io/druid/query/aggregation/datasketches/theta/SketchAggregatorFactory.java]: Reverted this change

**Comment by leventov** [comment]
@himanshug addressed comments

File: java-util/src/main/java/io/druid/java/util/common/StringUtils.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (himanshug) [reply java-util/src/main/java/io/druid/java/util/common/StringUtils.java]: ah... dint realize that. i'm surprised to know that there is no standard util to remove a character from a string :)

File: processing/src/main/java/io/druid/segment/DimensionSelector.java (hunk 1 above)
Comment (himanshug) [reply processing/src/main/java/io/druid/segment/DimensionSelector.java]: sounds good then

**Comment by himanshug** [comment]
@leventov given that this is incompatible and we haven't released 0.11.0 , is it possible to backport this to 0.11.0 ?

**Comment by leventov** [comment]
@himanshug thanks for review and merge, but this PR is tagged `Design Review` so it needed another one review from a committer

**Comment by leventov** [comment]
Anyway IMO it's too big and risky for backport at this point. I like the idea of making the next version 0.12 more, moreover I'm pretty sure there going to be more PRs that would break compatibility

**Comment by leventov** [comment]
E. g. somebody may want to implement #4888

**Comment by himanshug** [comment]
ok, next release gets to be 0.12.0 then. My apologies, I dint realize that you tagged it for additional reviews.

**Pull request merged**